File: native-android/app/src/main/java/de/ridetracker/session/RidePackage.kt
package de.ridetracker.session

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.security.MessageDigest
import java.time.Instant

object RidePackageStore {
    fun save(
        context: Context,
        session: RideSessionDocument,
        telemetryFile: File,
    ): File {
        val media = JSONArray()
        val hashes = JSONObject()
        hashes.put(telemetryFile.name, sha256(telemetryFile))

        session.videoFilename?.let { filename ->
            val videoFile = File(context.filesDir, filename)
            val mediaEntry = JSONObject()
                .put("kind", "video")
                .put("filename", filename)
                .put("sessionID", session.id)
                .put("startOffsetSeconds", session.videoStartOffsetSeconds)
                .put("mimeType", if (filename.endsWith(".mov", true)) "video/quicktime" else "video/mp4")
            if (videoFile.exists()) {
                mediaEntry.put("sizeBytes", videoFile.length())
                hashes.put(filename, sha256(videoFile))
            }
            media.put(mediaEntry)
        }

        val manifest = JSONObject()
            .put("packageVersion", "1.0.0")
            .put("sessionID", session.id)
            .put("createdAt", Instant.now().toString())
            .put("platform", "android")
            .put("telemetry", JSONObject()
                .put("filename", telemetryFile.name)
                .put("schemaVersion", "2.0.0")
                .put("sessionID", session.id)
                .put("sampleCount", session.summary.sampleCount)
                .put("durationSeconds", session.summary.durationSeconds))
            .put("media", media)
            .put("context", JSONObject()
                .put("parkID", JSONObject.NULL)
                .put("rideID", JSONObject.NULL)
                .put("parkName", JSONObject.NULL)
                .put("rideName", JSONObject.NULL))
            .put("privacy", JSONObject()
                .put("visibility", "private")
                .put("locationPrecision", "exact"))
            .put("integrity", JSONObject()
                .put("algorithm", "sha256")
                .put("files", hashes))

        val packageFile = File(context.filesDir, telemetryFile.name.removeSuffix(".ride.json") + ".ride-package.json")
        packageFile.writeText(manifest.toString(2))
        return packageFile
    }

    private fun sha256(file: File): String {
        val digest = MessageDigest.getInstance("SHA-256")
        file.inputStream().use { input ->
            val buffer = ByteArray(DEFAULT_BUFFER_SIZE)
            while (true) {
                val count = input.read(buffer)
                if (count <= 0) break
                digest.update(buffer, 0, count)
            }
        }
        return digest.digest().joinToString("") { "%02x".format(it) }
    }
}

File: native-ios/RideTrackerNative/RidePackage.swift
import CryptoKit
import Foundation

private struct RidePackageTelemetry: Codable {
    let filename: String
    let schemaVersion: String
    let sessionID: String
    let sampleCount: Int
    let durationSeconds: Double
}

private struct RidePackageMedia: Codable {
    let kind: String
    let filename: String
    let sessionID: String
    let startOffsetSeconds: Double
    let mimeType: String
    let sizeBytes: Int?
}

private struct RidePackagePrivacy: Codable {
    let visibility: String
    let locationPrecision: String
}

private struct RidePackageIntegrity: Codable {
    let algorithm: String
    let files: [String: String]
}

private struct RidePackageManifest: Codable {
    let packageVersion: String
    let sessionID: String
    let createdAt: Date
    let platform: String
    let telemetry: RidePackageTelemetry
    let media: [RidePackageMedia]
    let configurationSnapshot: CoreNativeConfigurationSnapshot?
    let privacy: RidePackagePrivacy
    let integrity: RidePackageIntegrity
}

enum RidePackageStore {
    static func save(session: RideSessionDocument, telemetryURL: URL) throws -> URL {
        var hashes = [telemetryURL.lastPathComponent: try sha256(telemetryURL)]
        var media: [RidePackageMedia] = []

        if let filename = session.video.filename {
            let videoURL = telemetryURL.deletingLastPathComponent().appendingPathComponent(filename)
            let exists = FileManager.default.fileExists(atPath: videoURL.path)
            if exists { hashes[filename] = try sha256(videoURL) }
            let size = exists
                ? (try? FileManager.default.attributesOfItem(atPath: videoURL.path)[.size] as? NSNumber)?.intValue
                : nil
            media.append(RidePackageMedia(
                kind: "video",
                filename: filename,
                sessionID: session.id.uuidString,
                startOffsetSeconds: session.video.startOffsetSeconds,
                mimeType: filename.lowercased().hasSuffix(".mov") ? "video/quicktime" : "video/mp4",
                sizeBytes: size
            ))
        }

        let manifest = RidePackageManifest(
            packageVersion: "1.0.0",
            sessionID: session.id.uuidString,
            createdAt: Date(),
            platform: "ios",
            telemetry: RidePackageTelemetry(
                filename: telemetryURL.lastPathComponent,
                schemaVersion: session.schemaVersion,
                sessionID: session.id.uuidString,
                sampleCount: session.summary.sampleCount,
                durationSeconds: session.summary.durationSeconds
            ),
            media: media,
            configurationSnapshot: session.configurationSnapshot,
            privacy: RidePackagePrivacy(visibility: "private", locationPrecision: "exact"),
            integrity: RidePackageIntegrity(algorithm: "sha256", files: hashes)
        )

        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        encoder.dateEncodingStrategy = .iso8601
        let data = try encoder.encode(manifest)
        let filename = telemetryURL.lastPathComponent.replacingOccurrences(of: ".ride.json", with: ".ride-package.json")
        let url = telemetryURL.deletingLastPathComponent().appendingPathComponent(filename)
        try data.write(to: url, options: .atomic)
        return url
    }

    private static func sha256(_ url: URL) throws -> String {
        let data = try Data(contentsOf: url)
        return SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
    }
}
